### PR TITLE
[WIP][Draft] Implement mergewith for bytesRange and bytesValue

### PR DESCRIPTION
**Pull request by atanu1991**
Addressed most comments. Two things remain
Merging ranges/byte values in MultiRange::mergeWith
Multiple ranges while merging bytes range with bytes range

### CODE REVIEW

**Comment by amaliujia** [comment]
Per this PR, seems to me that I will need to address `Merging ranges/byte values in MultiRange::mergeWith` in #299?

File: velox/type/Filter.cpp [hunk 1]
```diff
@@ -779,4 +779,240 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
 }
+
+std::unique_ptr<Filter> BytesRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull:
+      return std::make_unique<BytesRange>(*this, false);
```
Comment (amaliujia) [inline velox/type/Filter.cpp:790]: Nit: I will suggest to add /*nullAllowed=*/ for `false` argument (which will look like `return std::make_unique<BytesRange>(*this, /*nullAllowed=*/false);`

Basically this will increase the readability for a constant assignment of an argument. Same for other cases in this PR.

**Comment by amaliujia** [comment]
> Per this PR, seems to me that I will need to address `Merging ranges/byte values in MultiRange::mergeWith` in #299?

cc @mbasmanova  to confirm?

File: velox/type/tests/FilterBuilder.h [hunk 2]
```diff
@@ -227,6 +227,20 @@ inline std::unique_ptr<common::BytesRange> greaterThanOrEqual(
       min, false, false, "", true, true, nullAllowed);
 }
 
+inline std::unique_ptr<common::BytesRange> lessThan(
+    const std::string& max,
+    bool nullAllowed = false) {
+  return std::make_unique<common::BytesRange>(
```
Comment (mbasmanova) [inline velox/type/tests/FilterBuilder.h:233]: For <, upperExclusive should be true, right? Similar comment below.

**Comment by atanu1991** [comment]
> Per this PR, seems to me that I will need to address `Merging ranges/byte values in MultiRange::mergeWith` in #299?

I have addressed that here. Probably there would some changes required to consider merging bytes ranges and values based of your implementation in #299

File: velox/type/Filter.cpp [hunk 3]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
       VELOX_UNREACHABLE();
   }
 }
+
+namespace {
+bool mergeUpperExclusive(
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:798]: mergeUpperExclusive and mergeLowerExclusive are very similar. I wonder if these can be combined into a single function:

```
// compareResult = left < right for upper, right < left for lower
bool mergeExclusive(int compareResult, bool left, bool right) {
  return compareResult == 0 ? (left || right)
                            : (compareResult < 0 ? left : right);
}
```

File: velox/type/Filter.cpp [hunk 4]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
+      return std::make_unique<BytesRange>(*this, false);
+    case FilterKind::kBytesValues:
+    case FilterKind::kMultiRange:
+      return other->mergeWith(this);
+    case FilterKind::kBytesRange: {
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+
+      auto otherRange = static_cast<const BytesRange*>(other);
+
+      bool upperUnbounded = false, lowerUnbounded = false;
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:849]: nit: avoid multiple declarations; declare each variable separately.

File: velox/type/Filter.cpp [hunk 5]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
+
+      // TODO:
+      // Handle the case of same filter being upperUnbounded and
+      // lowerUnbounded at the same time. The output of this merge
+      // can result in multiple ranges.
+
+      assert(!(this->upperUnbounded_ && this->lowerUnbounded_));
+      assert(!(otherRange->upperUnbounded_ && otherRange->lowerUnbounded_));
+
+      if (this->upperUnbounded_ && otherRange->upperUnbounded_) {
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:862]: This is a bit long. I wonder if this logic can be shortened a bit:

```
      if (lowerUnbounded_) {
        lowerUnbounded = otherRange->lowerUnbounded_;
        lowerExclusive = otherRange->lowerExclusive_;
        lower = otherRange->lower_;
      } else if (otherRange->lowerUnbounded_) {
        lowerUnbounded = lowerUnbounded_;
        lowerExclusive = lowerExclusive_;
        lower = lower_;
      } else {
        lowerUnbounded = false;
        auto compare = lower_.compare(otherRange->lower_);
        lower = compare < 0 ? otherRange->lower_ : lower_;
        lowerExclusive = mergeExclusive(
            -compare, lowerExclusive_, otherRange->lowerExclusive_);
      }

      if (upperUnbounded_) {
        upperUnbounded = otherRange->upperUnbounded_;
        upperExclusive = otherRange->upperExclusive_;
        upper = otherRange->upper_;
      } else if (otherRange->upperUnbounded_) {
        upperUnbounded = upperUnbounded_;
        upperExclusive = upperExclusive_;
        upper = upper_;
      } else {
        upperUnbounded = false;
        auto compare = upper_.compare(otherRange->upper_);
        upper = compare < 0 ? upper_ : otherRange->upper_;
        upperExclusive = mergeExclusive(
            compare, upperExclusive_, otherRange->upperExclusive_);
      }

      if (!lowerUnbounded && !upperUnbounded &&
          (lower > upper ||
           (lower == upper && (lowerExclusive || upperExclusive)))) {
        return nullOrFalse(bothNullAllowed);
      }

      return std::make_unique<BytesRange>(
          lower,
          lowerUnbounded,
          lowerExclusive,
          upper,
          upperUnbounded,
          upperExclusive,
          bothNullAllowed);
```

File: velox/type/Filter.cpp [hunk 6]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
+      return std::make_unique<BytesValues>(
+          std::move(newValues), bothNullAllowed);
+    }
+
+    default:
+      VELOX_UNREACHABLE();
+  }
+}
+
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:1038]: Is this just a move? Are there any changes? It would be easier to review if this method stayed in place.

File: velox/type/Filter.h [hunk 7]
```diff
@@ -1262,6 +1269,16 @@ class MultiRange final : public Filter {
   std::unique_ptr<Filter> clone(
       std::optional<bool> nullAllowed = std::nullopt) const final;
 
+  const std::vector<std::unique_ptr<Filter>>& filters() const {
```
Comment (mbasmanova) [inline velox/type/Filter.h:1272]: Any particular reason for moving these methods? It is hard to review when moves are mixed with material changes.

File: velox/type/Filter.cpp [hunk 8]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
+
+std::unique_ptr<Filter> BytesValues::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+    case FilterKind::kMultiRange:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull:
+      return std::make_unique<BytesValues>(*this, false);
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:974]: nit: perhaps, return clone(false) ?

File: velox/type/Filter.cpp [hunk 9]
```diff
@@ -847,4 +793,330 @@ std::unique_ptr<Filter> BigintMultiRange::mergeWith(const Filter* other) const {
...
+      }
+
+      return std::make_unique<BytesValues>(
+          std::move(newValues), bothNullAllowed);
+    }
+    case FilterKind::kBytesRange: {
+      auto otherBytesRange = static_cast<const BytesRange*>(other);
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+
+      if ((!otherBytesRange->isLowerUnbounded() &&
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:1010]: nit: consider re-using testBytesRange:

`if (!testBytesRange(otherBytesRange->isLowerUnbounded() ? std::nullopt : otherBytesRange->lower()), otherBytesRange->isUpperUnbounded() ? std::nullopt : otherBytesRange->upper())`

File: velox/type/tests/FilterTest.cpp [hunk 10]
```diff
@@ -504,6 +504,34 @@ TEST(FilterTest, bytesRange) {
   EXPECT_FALSE(filter->testBytes("banana", 6));
   EXPECT_FALSE(filter->testBytes("camel", 5));
   EXPECT_FALSE(filter->testBytes("_abc", 4));
+
+  // < b
+  filter =
+      std::make_unique<BytesRange>("", true, false, "b", false, true, false);
```
Comment (mbasmanova) [inline velox/type/tests/FilterTest.cpp:510]: Any particular reason not to use the new lessThan("b") helper function?

File: velox/type/tests/FilterTest.cpp [hunk 11]
```diff
@@ -504,6 +504,34 @@ TEST(FilterTest, bytesRange) {
...
+  EXPECT_TRUE(filter->testBytes("a", 1));
+  EXPECT_FALSE(filter->testBytes("b", 1));
+  EXPECT_FALSE(filter->testBytes("c", 1));
+
+  // <= b
+  filter =
```
Comment (mbasmanova) [inline velox/type/tests/FilterTest.cpp:516]: lessThanOrEqual("b") ?

File: velox/type/tests/FilterTest.cpp [hunk 12]
```diff
@@ -504,6 +504,34 @@ TEST(FilterTest, bytesRange) {
...
+      std::make_unique<BytesRange>("", true, false, "b", false, false, false);
+  EXPECT_TRUE(filter->testBytes("a", 1));
+  EXPECT_TRUE(filter->testBytes("b", 1));
+  EXPECT_FALSE(filter->testBytes("c", 1));
+
+  // >= b
+  filter =
```
Comment (mbasmanova) [inline velox/type/tests/FilterTest.cpp:523]: ditto

File: velox/type/tests/FilterTest.cpp [hunk 13]
```diff
@@ -504,6 +504,34 @@ TEST(FilterTest, bytesRange) {
...
+      std::make_unique<BytesRange>("b", false, false, "", true, true, false);
+  EXPECT_FALSE(filter->testBytes("a", 1));
+  EXPECT_TRUE(filter->testBytes("b", 1));
+  EXPECT_TRUE(filter->testBytes("c", 1));
+
+  // > b
+  filter =
+      std::make_unique<BytesRange>("b", false, true, "", true, false, false);
```
Comment (mbasmanova) [inline velox/type/tests/FilterTest.cpp:531]: ditto

**Review by mbasmanova — COMMENTED** [review]
@atanu1991 Overall looks good. Some comments below.

Consider adding BytesRange::toString() method to help debugging test failures:

```
  std::string toString() const final {
    return fmt::format(
        "BytesRange: {}{}, {}{} {}",
        (lowerUnbounded_ || lowerExclusive_) ? "(" : "[",
        lowerUnbounded_ ? "..." : lower_,
        upperUnbounded_ ? "..." : upper_,
        (upperUnbounded_ || upperExclusive_) ? ")" : "]",
        nullAllowed_ ? "with nulls" : "no nulls");
  }
```

**Comment by mbasmanova** [comment]
There are some typos in the PR title and description.

Title: 

> Implement mergewith for bytesRange and bytesValue

Implement BytesRange::mergeWith and BytesValues::mergeWith

Description:

- "Implement mergewith for bytesRange and bytesValue" is redundant as it repeats the title. Let's remove.
- "and introduces mergeWith implmentation for BytesRange and BytesValues." - this is also redundant. It is just a different way to say what the title already says.

> (New PR) Handling case of merging multiple byte ranges which results in MultiRange as the output

Any particular reason not to include this functionality in this PR? Would this require significant refactoring of the current implementation?

File: velox/type/Filter.h (hunk 7 above)
Comment (atanu1991) [reply velox/type/Filter.h:1272]: They are not moves exactly. I had them in one place but #299 was pushed which included the same functions in a different location, hence it shows up this way on github

**Comment by atanu1991** [comment]
Thanks a lot for the detailed review.

@mbasmanova The case where the same range is both lower and upper unbounded might result in a MultiRange when merged with another range. This makes it a bit different from the existing cases now, and needs to be handled in a special way. Hence I thought handling it in a new PR would be better. Let me know your thoughts.

**Comment by mbasmanova** [comment]
> The case where the same range is both lower and upper unbounded might result in a MultiRange when merged with another range.

Can you give an example? How are you planning to handle this case?

**Comment by mbasmanova** [comment]
> the same range is both lower and upper unbounded

Is this possible? That's just kAlwaysTrue. We can add a check to the constructor to make sure if such range is never created.

**Comment by mbasmanova** [comment]
I see that comments in Filter.h are not explaining that MultiRange filter is expected to be a combination of non-overlapping filters and in general each filter must be restrictive, e.g. cannot pass all non-null values unless it is kAlwaysTrue or kIsNotNull. We need to clarify that.

**Comment by atanu1991** [comment]
> > The case where the same range is both lower and upper unbounded might result in a MultiRange when merged with another range.
> 
> Can you give an example? How are you planning to handle this case?

Example a range like (<=c, >=q) ("q", true, false "c", true, false) merged with [a-z]
The output of this would be [a-c] and [q-z] right?

I havent quite thought of exactly how to handle this right now. 
Was planning to have a special if block " if (upperunbounded && lowerunbounded) " and handle it there.

**Comment by mbasmanova** [comment]
> Example a range like (<=c, >=q) ("q", true, false "c", true, false) 

That would be a MultiRange, right? Not kBytesRange. The TODO is currently in the code that merges two BytesRange's. Is it in the right place?

File: velox/type/Filter.cpp [hunk 14]
```diff
@@ -490,35 +490,65 @@ std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
       return this->clone(/*nullAllowed=*/false);
     case FilterKind::kDoubleRange:
     case FilterKind::kFloatRange:
-    case FilterKind::kBytesRange:
-    case FilterKind::kBytesValues:
-      // TODO Implement
+      // TODO: Implement
```
Comment (atanu1991) [inline velox/type/Filter.cpp:493]: @mbasmanova  Added the function back to the right  location

**Comment by atanu1991** [comment]
> > Example a range like (<=c, >=q) ("q", true, false "c", true, false)
> 
> That would be a MultiRange, right? Not kBytesRange. The TODO is currently in the code that merges two BytesRange's. Is it in the right place?

I see. So basically is it impossible to create a BytesRange object with ("q", true, false "c", true, false)
If yes, then should we have that assert in the BytesRange Constructor?

And yes if this is a MultiRange then there is no TODO remaining for this PR

**Comment by mbasmanova** [comment]
BytesRange can only represent a single range. It cannot represent <= a OR >= b. Let's add a check to the constructor in a separate PR.

File: velox/type/Filter.cpp (hunk 14 above)
Comment (mbasmanova) [reply velox/type/Filter.cpp:493]: Thanks.

File: velox/type/Filter.cpp [hunk 15]
```diff
@@ -490,35 +490,65 @@ std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
...
       VELOX_UNREACHABLE();
+    case FilterKind::kBytesValues:
+    case FilterKind::kBytesRange:
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:496]: I find this logic very hard to follow. Let's split kBytesXxx case from kMultiRange. We can then see if common logic can be factored out into helper methods.

Comment (atanu1991) [reply velox/type/Filter.cpp:496]: Ah ok. The logic is exactly the same.

What I do here is if kMultiRange I extract all the filters in a vector, otherwise the new vector otherFilters has just one element.

Post that its the same logic from line 517-551

Comment (mbasmanova) [reply velox/type/Filter.cpp:496]: I wonder if it would be simpler if BytesValues::mergeWith implemented merging with kMultiRange directly. The result would always be a kBytesValues filter (or kAlwaysFalse or kIsNull), right?

File: velox/type/Filter.cpp [hunk 16]
```diff
@@ -490,35 +490,65 @@ std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
...
     case FilterKind::kMultiRange: {
-      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
       bool bothNullAllowed = nullAllowed_ && other->testNull();
-      bool bothNanAllowed = nanAllowed_ && multiRangeOther->nanAllowed_;
+      bool bothNanAllowed = nanAllowed_;
+
+      std::vector<const Filter*> otherFilters;
       std::vector<std::unique_ptr<Filter>> merged;
```
Comment (mbasmanova) [inline velox/type/Filter.cpp:502]: nit: Let's move these variables closer to where they are used.

File: velox/type/Filter.cpp (hunk 15 above)
Comment (mbasmanova) [reply velox/type/Filter.cpp:496]: I think I understand how this works now.

**Review by mbasmanova — APPROVED** [review]
Looks good to me. Thank you for the contribution.

File: velox/type/Filter.cpp (hunk 15 above)
Comment (atanu1991) [reply velox/type/Filter.cpp:496]: Thats right. BytesValues::mergeWith implemented merging with kMultiRange will always return a BytesValues. Should I go ahead and make that change too?

Comment (mbasmanova) [reply velox/type/Filter.cpp:496]: Up to you. Whatever you think is best.

Comment (atanu1991) [reply velox/type/Filter.cpp:496]: I will leave it here. I think the code doesnt duplicate if we do this.

Will address the nits, and create the fbsource diff and land this PR.

Thanks!
